### ISSUE TITLE
[cfe] Type `void` is allowed in `[]=` operator

### ISSUE THREAD

**Issue by sgrekhov** [dart-lang/sdk#57071]
The code below produce no issues in the analyzer but errors in CFE.
```dart
main() {
  Map<void, int> m1 = <void, int>{};
  void key = 0;
  m1[key] = 1; // Error: This expression has type 'void' and can't be used.

  Map<int, void> m2 = <int, void>{};
  void value = 0;
  m2[2] = value; // Error: This expression has type 'void' and can't be used.
}
```
Acording to the [specification](https://github.com/dart-lang/language/blob/a530cd6dac9c43f8cdd91aa95a58ddf4e0d9c463/specification/dartLangSpec.tex#L22755-L22777) key and value of type `void` in the `[]=` operator are both allowed.

See also #57070.

**Comment by eernstg** [comment]
Right. It could be argued that even though `operator []=` is a method with two parameters, the syntax `m1[key] = 1` is so different from the regular `myFunction(arg1, arg2)` that it should be given special treatment.

(For example, `e1 == e2` is not treated exactly the same way as an invocation of a method with two arguments.)

However, I don't think there is any particular reason why we would want to make an exception for this case, so I'd recommend that we consider the spec rule ["an actual argument of type `void` can be passed to a parameter of type `void`"](https://github.com/dart-lang/language/blob/a530cd6dac9c43f8cdd91aa95a58ddf4e0d9c463/specification/dartLangSpec.tex#L22734) to be applicable to this case as well.

**Comment by lrhn** [comment]
Having a non-empty map with `void` as key type is a _bad_ idea to begin with.
Following the rule that a `void` typed expression may occur in a `void`-typed context, this does seem like it should be allowed.
(I never liked that rule to begin with, so I'd be fine with exceptions, but the it would be the consistent specification.)

The spec phrasing could use a little tweaking:
> ```latex
>   If $s$ is an \synt{assignableSelector} of the
>   form \code{.\id} or \code{?.\id}
>   where the static type of the identifier \id{} is \VOID,
>   $e_1$ may have type \VOID.
>   Otherwise, if $s$ is an \synt{assignableSelector} of the form
>   \code{[$\,e_0\,$]} where the static type of
>   the first formal parameter in the statically known declaration
>   of operator \code{[]=} is \VOID,
>   $e_0$ may have type \VOID.
>   Also, if the static type of the second formal parameter is \VOID,
>   $e_1$ may have type \VOID.
> ```

The "where the static type of the identifier \\id{} is \\VOID" should probably say 
"where the parameter type of the statically know setter denoted by \\id{} is \\VOID". The identifier is not an expression, it doesn't have a static type. It's a member accessor on an expression which does have a static type, and it's invoking that type's *setter* for the `id` property. (Even if we allow ourselves to use a shorthand like "The type of the `id` identifier of type `T`", that would be more likely to refer to the getter return type, and is at least ambiguous when the types can differ between getter and setter.)

(_Ceterum autem censeo_ that `void` expressions should not be allowed anywhere where their value is not immediately discarded.)

**Comment by eernstg** [comment]
I think we'd then expect the following to succeed:
```dart
class IntVoid {
  void operator []=(int _, void _) {}
}

class VoidInt {
  void operator []=(void _, int _) {}
}

class VoidVoid {
  void operator []=(void _, void _) {}
}

void main() {
  void v = null;
  IntVoid()[0] = v;
  VoidInt()[v] = 0;
  VoidVoid()[v] = v;
}
```

Currently, this succeeds with the analyzer, but the CFE reports that `v` has a value that can't be used, at all four occurrences. Note that the following succeeds with the analyzer as well as the CFE:

```dart
set setter(void v) {}

void main() {
  void v = null;
  setter = v;
}
```

This is consistent with the perspective that the right hand side of an assignment that invokes a setter is treated as an actual argument to the formal parameter of said setter, which would allow both `e2` and `e3` to have type `void` in an expression of the form `e1[e2] = e3` when the corresponding formal parameters have type `void`.

**Comment by lrhn** [comment]
ACK, and similarly, the expression inside the `[..]` of the `[]` and `[]=` operators is also a parameter, and should be treated like one. The second paramter to `[]=` can be treated as a setter or a parameter, but it should give the same result. (Allowing `void` in a position where its value gets propagated ... yuck!)

**Comment by eernstg** [comment]
Let's get this one settled: https://github.com/dart-lang/language/pull/4454.

**Comment by sgrekhov** [comment]
See also https://github.com/dart-lang/sdk/issues/62844
